packages/fhir-router/src/batch.ts

```diff
@@ -99,86 +98,239 @@ class BatchProcessor {
+        parseSearchRequest(entry.request.url + '?' + entry.request.ifNoneExist)
+      );
+      if (existing.length === 1 && entry.fullUrl?.startsWith(uuidUriPrefix)) {
+        return { placeholder: entry.fullUrl, reference: getReferenceString(existing[0]) };
+      }
+    }
+    if (entry.resource && entry.fullUrl?.startsWith(uuidUriPrefix)) {
+      entry.resource.id = this.repo.generateId();
+      return {
+        placeholder: entry.fullUrl,
+        reference: getReferenceString(entry.resource),
+      };
+    }
+    return undefined;
   }
 
-  private validateEntry(entry: BundleEntry): void {
-    if (!entry.request) {
-      throw new OperationOutcomeError(badRequest('Missing entry.request'));
+  private async processModification(entry: BundleEntry, path: string): Promise<BundleEntryIdentity | undefined> {
+    if (entry.request?.url?.includes('?')) {
+      const method = entry.request.method;
+      // Resolve conditional update via search
+      const resolved = await this.repo.searchResources(parseSearchRequest(entry.request.url));
+      if (resolved.length !== 1) {
+        if (resolved.length === 0 && method === 'DELETE') {
+          return undefined;
+        }
+        if (resolved.length === 0 && method === 'PUT' && !entry.resource?.id) {
+          return undefined; // create by update for conditional PUT
+        }
+        throw new OperationOutcomeError(
+          badRequest(`Conditional ${entry.request.method} matched ${resolved.length} resources`, path + '.request.url')
+        );
+      }
+      const reference = getReferenceString(resolved[0]);
+      entry.request.url = reference;
+      return { placeholder: entry.request.url, reference };
+    } else if (entry.request?.url.includes('/')) {
+      return { placeholder: entry.request.url, reference: entry.request.url };
     }
+    return undefined;
+  }
 
-    if (!entry.request.method) {
-      throw new OperationOutcomeError(badRequest('Missing entry.request.method'));
+  /**
+   * Processes a single entry from a FHIR batch request.
+   * @param entry - The bundle entry.
+   * @returns The bundle entry response.
+   */
+  private async processBatchEntry(entry: BundleEntry): Promise<BundleEntry> {
+    const [outcome, resource] = await this.performBatchOperation(entry);
+    if (!isOk(outcome) && this.bundle.type === 'transaction') {
+      throw new OperationOutcomeError(outcome);
     }
+    return buildBundleResponse(outcome, resource);
+  }
 
-    if (!entry.request.url) {
-      throw new OperationOutcomeError(badRequest('Missing entry.request.url'));
+  private async performBatchOperation(entry: BundleEntry): Promise<FhirResponse> {
+    const urlParts = splitN(entry.request?.url as string, '?', 2);
+    const requestPath = urlParts[0];
+    const queryParams = urlParts[1];
+    const route = this.router.find(entry.request?.method as HttpMethod, requestPath);
+    const params = route?.params;
+
+    switch (route?.data?.interaction) {
```

Comment:
Hm, I need to review this more, but I would have assumed that most of this `switch` statement would have been handled by `handleRequest` - if nothing else, a lot of these checks are duplicated, which would be nice to clean up if possible.

```ts
    const result = await this.router.handleRequest(
      {
        method: request.method as HttpMethod,
        pathname: url.pathname,
        params: Object.create(null),
        query: Object.fromEntries(url.searchParams),
        body,
      },
      this.repo
    );
```